### PR TITLE
Add `RubyEvent#dup` support and unit test case to keep `Json#dump(Event)` safe.

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

[rn:skip]

## What does this PR do?
Covers two areas:
- Adds a support `RubyEvent#dup` additional to `RubyEvent#clone`. It is a safety guard when we use duplicate `dup` methods on the Logstash event object;
- Without `RubyEvent#dup` we got `NullPointerException`. To make sure we don't break `Json#dump` which internally duplicated the event before (it wasn't necessary so we removed) on any future changes, this PR also adds a small unit test.

`JrubyEventExtLibrary.java` changes are code formatting (couldn't close my 👁️ ) except 157L.
Exhaustive CI run: https://buildkite.com/elastic/logstash-exhaustive-tests-pipeline/builds/551

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

File: logstash-core/spec/logstash/event_spec.rb [hunk 1]
```diff
@@ -428,5 +428,32 @@
       expect(cloned.get("[@metadata][fancy]")).to eq("foobar")
       expect(event1.get("[@metadata][fancy]")).to eq("pants")
     end
+
+    it "validates that cloned and duped event contents are same" do
```
Comment (yaauie) [inline logstash-core/spec/logstash/event_spec.rb:432]: If we expect `dup` to be an alias of `clone`, then we should apply all of the specs to it:

~~~ diff
diff --git a/logstash-core/spec/logstash/event_spec.rb b/logstash-core/spec/logstash/event_spec.rb
index d879100c7..a21a84f05 100644
--- a/logstash-core/spec/logstash/event_spec.rb
+++ b/logstash-core/spec/logstash/event_spec.rb
@@ -373,60 +373,76 @@ describe LogStash::Event do
     end
   end
 
-  describe "#clone" do
-    let(:fieldref) { "[@metadata][fancy]" }
-    let(:event1) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => "pants" }) }
-    let(:event2) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => "pants2"} }) }
-    let(:event3) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => {"fancy3" => "pants2"}} }) }
-    let(:event4) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => ["pants1", "pants2"]} }) }
-    let(:event5) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => "pants", "smarty" => "pants2" }) }
+  %i(dup clone).each do |clone_method|
+    describe "##{clone_method}" do
+
+      let(:fieldref) { "[@metadata][fancy]" }
+      let(:event1) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => "pants" }) }
+      let(:event2) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => "pants2"} }) }
+      let(:event3) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => {"fancy3" => "pants2"}} }) }
+      let(:event4) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => {"fancy2" => ["pants1", "pants2"]} }) }
+      let(:event5) { LogStash::Event.new("hello" => "world", "@metadata" => { "fancy" => "pants", "smarty" => "pants2" }) }
+
+      it "should clone metadata fields" do
+        cloned = event.public_send(clone_method)
+        expect(cloned.get(fieldref)).to eq("pants")
+        expect(cloned.to_hash_with_metadata).to include("@metadata")
+      end
 
-    it "should clone metadata fields" do
-      cloned = event1.clone
-      expect(cloned.get(fieldref)).to eq("pants")
-      expect(cloned.to_hash_with_metadata).to include("@metadata")
-    end
+      it "should clone metadata fields with nested json" do
+        cloned = event2.public_send(clone_method)
+        expect(cloned.get(fieldref)).to eq({"fancy2" => "pants2"})
+        expect(cloned.get("hello")).to eq("world")
+        expect(cloned.to_hash).not_to include("@metadata")
+        expect(cloned.to_hash_with_metadata).to include("@metadata")
+      end
 
-    it "should clone metadata fields with nested json" do
-      cloned = event2.clone
-      expect(cloned.get(fieldref)).to eq({"fancy2" => "pants2"})
-      expect(cloned.get("hello")).to eq("world")
-      expect(cloned.to_hash).not_to include("@metadata")
-      expect(cloned.to_hash_with_metadata).to include("@metadata")
-    end
+      it "should clone metadata fields with 2-level nested json" do
+        cloned = event3.public_send(clone_method)
+        expect(cloned.get(fieldref)).to eq({"fancy2" => {"fancy3" => "pants2"}})
+        expect(cloned.to_hash).not_to include("@metadata")
+        expect(cloned.to_hash_with_metadata).to include("@metadata")
+      end
 
-    it "should clone metadata fields with 2-level nested json" do
-      cloned = event3.clone
-      expect(cloned.get(fieldref)).to eq({"fancy2" => {"fancy3" => "pants2"}})
-      expect(cloned.to_hash).not_to include("@metadata")
-      expect(cloned.to_hash_with_metadata).to include("@metadata")
-    end
+      it "should clone metadata fields with nested json and array value" do
+        cloned = event4.public_send(clone_method)
+        expect(cloned.get(fieldref)).to eq({"fancy2" => ["pants1", "pants2"]})
+        expect(cloned.to_hash_with_metadata).to include("@metadata")
+      end
 
-    it "should clone metadata fields with nested json and array value" do
-      cloned = event4.clone
-      expect(cloned.get(fieldref)).to eq({"fancy2" => ["pants1", "pants2"]})
-      expect(cloned.to_hash_with_metadata).to include("@metadata")
-    end
+      it "should clone metadata fields with multiple keys" do
+        cloned = event5.public_send(clone_method)
+        expect(cloned.get(fieldref)).to eq("pants")
+        expect(cloned.get("[@metadata][smarty]")).to eq("pants2")
+        expect(cloned.to_hash_with_metadata).to include("@metadata")
+      end
 
-    it "should clone metadata fields with multiple keys" do
-      cloned = event5.clone
-      expect(cloned.get(fieldref)).to eq("pants")
-      expect(cloned.get("[@metadata][smarty]")).to eq("pants2")
-      expect(cloned.to_hash_with_metadata).to include("@metadata")
-    end
+      it "mutating cloned event should not affect the original event" do
+        cloned = event1.public_send(clone_method)
+        cloned.set("hello", "foobar")
+        expect(cloned.get("hello")).to eq("foobar")
+        expect(event1.get("hello")).to eq("world")
+      end
 
-    it "mutating cloned event should not affect the original event" do
-      cloned = event1.clone
-      cloned.set("hello", "foobar")
-      expect(cloned.get("hello")).to eq("foobar")
-      expect(event1.get("hello")).to eq("world")
-    end
+      it "mutating cloned event's metadata should not affect the original event metadata" do
+        cloned = event1.public_send(clone_method)
+        cloned.set("[@metadata][fancy]", "foobar")
+        expect(cloned.get("[@metadata][fancy]")).to eq("foobar")
+        expect(event1.get("[@metadata][fancy]")).to eq("pants")
+      end
+
+      it "returns a new clone each time" do
+        cloneOne = event1.public_send(clone_method)
+        cloneTwo = event1.public_send(clone_method)
+        cloneOne.set("[@metadata][fancy]", "dress")
+        cloneTwo.set("[@metadata][fancy]", "dinner")
 
-    it "mutating cloned event's metadata should not affect the original event metadata" do
-      cloned = event1.clone
-      cloned.set("[@metadata][fancy]", "foobar")
-      expect(cloned.get("[@metadata][fancy]")).to eq("foobar")
-      expect(event1.get("[@metadata][fancy]")).to eq("pants")
+        expect(cloneTwo).to_not be cloneOne
+
+        expect(event1.get("[@metadata][fancy]")).to eq("pants")
+        expect(cloneOne.get("[@metadata][fancy]")).to eq("dress")
+        expect(cloneTwo.get("[@metadata][fancy]")).to eq("dinner")
+      end
     end
   end
 end
~~~

File: logstash-core/spec/logstash/event_spec.rb [hunk 2]
```diff
@@ -428,5 +428,32 @@
...
+        [sample_event, dumped_sample_event]
+      end
+    end
+
+    it 'validates the Json#dump event' do
+      sample_events.each do |sample_event, dumped_sample_event|
+        # Json#dump internally normalizes Event as it is
+        expect(dumped_sample_event).eql?(sample_event)
+      end
+    end
```
Comment (yaauie) [inline logstash-core/spec/logstash/event_spec.rb:457]: I _believe_ this spec belongs with the json implementation

File: logstash-core/spec/logstash/event_spec.rb (hunk 1 above)
Comment (mashhurs) [reply logstash-core/spec/logstash/event_spec.rb]: This suggestion is great and I have applied. Thanks alot!

**Review by yaauie — APPROVED** [review]
👍🏼

**Pull request merged**